cmd/kubeadm/app/master/selfhosted.go

```diff
@@ -69,9 +87,271 @@ func CreateSelfHostedControlPlane(cfg *kubeadmapi.MasterConfiguration, client *c
+}
+
+func CreateEtcdCluster(cfg *kubeadmapi.MasterConfiguration, client *clientset.Clientset) error {
+	start := time.Now()
+
+	// setup TPR client
+	restClient, err := getEtcdTPRClient()
+	if err != nil {
+		return err
+	}
+
+	fmt.Println("Wait for TPR to exist")
+	k8sutil.WaitEtcdTPRReady(restClient, time.Second*5, time.Minute*1, "kube-system")
+
+	seedPodIP, err := getBootEtcdPodIP(client)
+	if err != nil {
+		return err
+	}
+	fmt.Printf("Boot IP is %s\n", seedPodIP)
+
+	// create TPR data
+	cluster := &spec.Cluster{
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: "etcd.coreos.com/v1beta1",
+			Kind:       "Cluster",
+		},
+		Metadata: metav1.ObjectMeta{
+			Name:      etcdCluster,
+			Namespace: metav1.NamespaceSystem,
+		},
+		Spec: spec.ClusterSpec{
+			Size:    1,
+			Version: "v3.1.6",
+			SelfHosted: &spec.SelfHostedPolicy{
+				BootMemberClientEndpoint: fmt.Sprintf("http://%s:12379", seedPodIP),
+			},
+			Pod: &spec.PodPolicy{
+				NodeSelector: map[string]string{"node-role.kubernetes.io/master": ""},
+				Tolerations: []v1.Toleration{
+					v1.Toleration{
+						Key:      "node-role.kubernetes.io/master",
+						Operator: "Exists",
+						Effect:   v1.TaintEffectNoSchedule,
+					},
+				},
+			},
+		},
+	}
+
+	fmt.Println("Sending TPR cluster data")
+	err = restClient.Post().
+		Resource(spec.TPRKindPlural).
+		Namespace(metav1.NamespaceSystem).
+		Body(cluster).
+		Do().Error()
+	if err != nil {
+		return err
+	}
+
+	fmt.Println("Waiting for 30s")
+	time.Sleep(30 * time.Second)
+
+	fmt.Println("Waiting for TPR data to exist")
+	err = wait.Poll(kubeadmconstants.DiscoveryRetryInterval, 5*time.Minute, func() (bool, error) {
+		cluster := &spec.Cluster{}
+
+		err := restClient.Get().
+			Resource(spec.TPRKindPlural).
+			Namespace(metav1.NamespaceSystem).
+			Name(etcdCluster).
+			Do().Into(cluster)
+
+		if err != nil {
+			fmt.Printf("[self-hosted] Error retrieving etcd cluster: %v\n", err)
+			return false, nil
+		}
+
+		switch cluster.Status.Phase {
+		case spec.ClusterPhaseRunning:
+			return true, nil
+		case spec.ClusterPhaseFailed:
+			return false, errors.New("failed to create etcd cluster")
+		default:
+			return false, nil
+		}
+	})
+	if err != nil {
+		return err
+	}
+
+	fmt.Println("Waiting for etcd to remove peer")
+	err = waitBootEtcdRemoved(cfg.Etcd.Cluster.ServiceIP)
+	if err != nil {
+		return err
+	}
+
+	fmt.Println("Removing seed etcd")
+	// remove seed etcd pod
+	etcdStaticManifestPath := buildStaticManifestFilepath(etcd)
+	if err := os.RemoveAll(etcdStaticManifestPath); err != nil {
+		return fmt.Errorf("unable to delete seed etcd manifest [%v]", err)
+	}
+
+	fmt.Println("Waiting for seed etcd pod to delete")
+
+	//wait for seed etcd to disappear
+	wait.PollInfinite(kubeadmconstants.DiscoveryRetryInterval, func() (bool, error) {
+		_, err := client.Core().Pods(metav1.NamespaceSystem).Get(etcd, metav1.GetOptions{})
+		if err != nil && apierrors.IsNotFound(err) {
+			return true, nil
+		}
+		return false, err
```

Comment:
If we want to error out in this case then you need to check the error returned by PollInfinite. Otherwise, just log this error.